sound/soc/intel/boards/skl_hda_dsp_generic.c

```diff
@@ -135,13 +144,19 @@ static int skl_hda_fill_card_info(struct snd_soc_acpi_mach_params *mach_params)
 				skl_hda_be_dai_links[IDISP_DAI_COUNT +
 					HDAC_DAI_COUNT + i];
 		}
-	} else if (codec_count == 2 && codec_mask & IDISP_CODEC_MASK) {
+	} else {
+		/* topology with external and iDisp HDA codecs */
 		num_links = ARRAY_SIZE(skl_hda_be_dai_links);
 		num_route = ARRAY_SIZE(skl_hda_map);
 		card->dapm_widgets = skl_hda_widgets;
 		card->num_dapm_widgets = ARRAY_SIZE(skl_hda_widgets);
-	} else {
-		return -EINVAL;
+		if (!idisp_mask) {
```

Comment:
@kv2019i Im  abit confused by this check. How will we get here? We already check for codec_count ==2 && idisp_mask on line 128, right?

Reply:
@ranj063 L128 checks for checks where system has two external codecs: "codec_count == 2 && !idisp_mask". So we reach this block when there are 1 codec and it is an external one.

Reply:
so this is where we take care of the case where the iDisp codecs are missing but the topology still has the idisp definitions?

Reply:
and looking at the second commit, the case when codec count is 1 and it is the HDA codec, we should choose the non-idisp topology no? So do we still need the dummy_codec assignments here?

Reply:
@ranj063 yes -- we basicly have two major DSP topology types for generic HDA. The first if branch is for the idisp-only topology, while the else branch handles the other topology.

Reply:
ahh so there is no topology that handles just the external HDA codec and no idisp codecs. gotcha!. sorry I got confused there for a bit.

Reply:
@ranj063 for your second question, yes, it is still needed. we do not have a "external codec only" topology for HDA. so in the case system has only one external codec, we need to use the topology that has idisp nodes, and it goes to the second branch in the machine driver (with dummy codec assigned)

I think we could use this approach in other cases as well to keep the number of distinct topology files reasonable. E.g. define the superset of connections in topology file, and at runtime disable graphs edges that are not connected.

Reply:
got it! thanks @kv2019i I think the PR looks good now!